src/Framework/ExecutionOrderDependency.php

```diff
@@ -0,0 +1,184 @@
+            $target      = $parts[0];
+        } else {
+            $cloneOption = $parts[0];
+            $target      = $parts[1];
+        }
+
+        // Prefix provided class for targets assumed to be in scope
+        if ($target !== '' && strpos($target, '::') === false) {
+            $target = $className . '::' . $target;
+        }
+
+        return new self($target, null, $cloneOption);
+    }
+
+    /**
+     * @param array<ExecutionOrderDependency> $dependencies
+     *
+     * @return array<ExecutionOrderDependency>
+     */
+    public static function filterInvalid(array $dependencies): array
+    {
+        return array_filter($dependencies, function (self $d) {
+            return $d->isValid();
+        });
+    }
+
+    /**
+     * @param array<ExecutionOrderDependency> $existing
+     * @param array<ExecutionOrderDependency  $additional
+     *
+     * @return array<ExecutionOrderDependency>
+     */
+    public static function mergeUnique(array $existing, array $additional): array
+    {
+        $existingTargets = array_map(function ($dependency) {
+            return $dependency->getTarget();
+        }, $existing);
+
+        foreach ($additional as $dependency) {
+            if (array_search($dependency->getTarget(), $existingTargets, true) === false) {
+                $existingTargets[]  = $dependency->getTarget();
+                $existing[]         = $dependency;
+            }
+        }
+
+        return $existing;
+    }
+
+    /**
+     * @param array<ExecutionOrderDependency> $left
+     * @param array<ExecutionOrderDependency  $right
+     *
+     * @return array<ExecutionOrderDependency>
+     */
+    public static function diff(array $left, array $right): array
+    {
+        if ($right === []) {
+            return $left;
+        }
+
+        if ($left === []) {
+            return [];
+        }
+
+        $diff         = [];
+        $rightTargets = array_map(function ($dependency) {
+            return $dependency->getTarget();
+        }, $right);
+
+        foreach ($left as $dependency) {
+            if (array_search($dependency->getTarget(), $rightTargets, true) === false) {
+                $diff[] = $dependency;
+            }
+        }
+
+        return $diff;
+    }
+
+    public function __construct(string $classOrCallableName, ?string $methodName = null, ?string $option = null)
+    {
+        if ($classOrCallableName === '') {
+            return $this;
+        }
+
+        if ($methodName !== null && $methodName !== '') {
+            $this->className  = $classOrCallableName;
+            $this->methodName = $methodName;
+        } elseif (strpos($classOrCallableName, '::') !== false) {
+            [$this->className, $this->methodName] = explode('::', $classOrCallableName);
+        } else {
+            $this->className  = $classOrCallableName;
+            $this->methodName = 'class';
+        }
+
+        if ($option === 'clone') {
+            $this->useDeepClone  = true;
```

Comment:
are these two booleans mutually exclusive? otherwise, it might be better to make an enum-type of construct using a string. Two separate booleans give the impression that they can be true simultaneously

Reply:
I agree that this looks strange. Honestly I am not sure about the functionality here, as I have never used it myself (or even seen it used by client projects).

Will refactor this further in the coming weeks as I gain more insight into this part of the `TestRunner`